File: 07-2.rb
#!/usr/bin/env ruby

class Hand
  attr_accessor :cards, :bid, :type

  include Comparable

  def initialize(cards, bid)
    @cards = cards.chars
    @bid = bid
    @type = classify
  end

  @@type_order = [:card, :one_pair, :two_pair, :three, :full_house, :four, :five]
  @@card_order = ['J', '2', '3', '4', '5', '6', '7', '8', '9', 'T', 'Q', 'K', 'A']

  def classify
    non_jokers = cards.select { |c| c != 'J' }.sort
    num_jokers = cards.count { |c| c == 'J' }

    case num_jokers
    when 5
      :five
    when 4
      :five
    when 3
      if non_jokers[0] == non_jokers[1]
        :five
      else
        :four
      end
    when 2
      if non_jokers[0] == non_jokers[1] and non_jokers[1] == non_jokers[2]
        :five
      elsif non_jokers[0] == non_jokers[1] or non_jokers[1] == non_jokers[2]
        :four
      else
        :three
      end
    when 1
      if non_jokers.all? { |c| c == non_jokers[0] }
        :five
      elsif non_jokers[0..2].all? { |c| c == non_jokers[0] } or
           non_jokers[1..3].all? { |c| c == non_jokers[1] }
        :four
      elsif non_jokers[0] == non_jokers[1] and non_jokers[2] == non_jokers[3]
        :full_house
      elsif non_jokers[0] == non_jokers[1] or
           non_jokers[1] == non_jokers[2] or
           non_jokers[2] == non_jokers[3]
        :three
      else
        :one_pair
      end
    when 0
      if non_jokers.all? { |c| c == non_jokers[0] }
        :five
      elsif non_jokers[0..-2].all? { |c| c == non_jokers[0] } or
           non_jokers[1..-1].all? { |c| c == non_jokers[1] }
        :four
      elsif (non_jokers[0] == non_jokers[1] and non_jokers[2] == non_jokers[3] and non_jokers[3] == non_jokers[4]) or
           (non_jokers[0] == non_jokers[1] and non_jokers[1] == non_jokers[2] and non_jokers[3] == non_jokers[4])
        :full_house
      elsif non_jokers[0..-3].all? { |c| c == non_jokers[0] } or
           non_jokers[1..-2].all? { |c| c == non_jokers[1] } or
           non_jokers[2..-1].all? { |c| c == non_jokers[2] }
        :three
      elsif (non_jokers[0] == non_jokers[1] and non_jokers[2] == non_jokers[3]) or
           (non_jokers[0] == non_jokers[1] and non_jokers[3] == non_jokers[4]) or
           (non_jokers[1] == non_jokers[2] and non_jokers[3] == non_jokers[4])
        :two_pair
      elsif non_jokers[0] == non_jokers[1] or
           non_jokers[1] == non_jokers[2] or
           non_jokers[2] == non_jokers[3] or
           non_jokers[3] == non_jokers[4]
        :one_pair
      else
        :card
      end
    end
  end

  def <=>(other)
    return -1 if @@type_order.index(@type) < @@type_order.index(other.type)
    return  1 if @@type_order.index(@type) > @@type_order.index(other.type)

    @cards.each_index do |n|
      return -1 if @@card_order.index(@cards[n]) < @@card_order.index(other.cards[n])
      return  1 if @@card_order.index(@cards[n]) > @@card_order.index(other.cards[n])
    end

    0
  end

  def inspect
    to_s
  end

  def to_s
    "<#{self.class}: #{@cards.join}, #{@bid}: #{@type}>"
  end
end

input = File.read('07.input').lines.map(&:strip)

hands = []
input.each do |line|
  hands << Hand.new(line.split[0], line.split[1].to_i)
end

hands.sort!

winnings = 0
hands.each_with_index do |h, i|
  winnings += h.bid * (i + 1)
end

print winnings, "\n"

File: 07-2.py
#!/usr/bin/env python

class Hand:
    type_order = ['card', 'one_pair', 'two_pair', 'three', 'full_house', 'four', 'five']
    card_order = ['J', '2', '3', '4', '5', '6', '7', '8', '9', 'T', 'Q', 'K', 'A']

    def __init__(self, cards, bid):
        self.cards = cards
        self.bid = bid
        self.hand_type = self.classify()

    def classify(self):
        non_jokers = sorted([c for c in self.cards if c != 'J'])
        num_jokers = len([c for c in self.cards if c == 'J'])

        match num_jokers:
            case 5:
                return 'five'
            case 4:
                return 'five'
            case 3:
                if non_jokers[0] == non_jokers[1]:
                    return 'five'
                else:
                    return 'four'
            case 2:
                if non_jokers[0] == non_jokers[1] and non_jokers[1] == non_jokers[2]:
                    return 'five'
                elif non_jokers[0] == non_jokers[1] or non_jokers[1] == non_jokers[2]:
                    return 'four'
                else:
                    return 'three'
            case 1:
                if non_jokers[0] == non_jokers[1] == non_jokers[2] == non_jokers[3]:
                    return 'five'
                elif non_jokers[0] == non_jokers[1] == non_jokers[2] or \
                     non_jokers[1] == non_jokers[2] == non_jokers[3]:
                    return 'four'
                elif non_jokers[0] == non_jokers[1] and non_jokers[2] == non_jokers[3]:
                    return 'full_house'
                elif non_jokers[0] == non_jokers[1] or \
                     non_jokers[1] == non_jokers[2] or \
                     non_jokers[2] == non_jokers[3]:
                    return 'three'
                else:
                    return 'one_pair'
            case 0:
                if non_jokers[0] == non_jokers[1] == non_jokers[2] == non_jokers[3] == non_jokers[4]:
                    return 'five'
                elif non_jokers[0] == non_jokers[1] == non_jokers[2] == non_jokers[3] or \
                     non_jokers[1] == non_jokers[2] == non_jokers[3] == non_jokers[4]:
                    return 'four'
                elif (non_jokers[0] == non_jokers[1] and non_jokers[2] == non_jokers[3] and non_jokers[3] == non_jokers[4]) or \
                     (non_jokers[0] == non_jokers[1] and non_jokers[1] == non_jokers[2] and non_jokers[3] == non_jokers[4]):
                    return 'full_house'
                elif non_jokers[0] == non_jokers[1] == non_jokers[2] or \
                     non_jokers[1] == non_jokers[2] == non_jokers[3] or \
                     non_jokers[2] == non_jokers[3] == non_jokers[4]:
                    return 'three'
                elif (non_jokers[0] == non_jokers[1] and non_jokers[2] == non_jokers[3]) or \
                     (non_jokers[0] == non_jokers[1] and non_jokers[3] == non_jokers[4]) or \
                     (non_jokers[1] == non_jokers[2] and non_jokers[3] == non_jokers[4]):
                    return 'two_pair'
                elif non_jokers[0] == non_jokers[1] or \
                     non_jokers[1] == non_jokers[2] or \
                     non_jokers[2] == non_jokers[3] or \
                     non_jokers[3] == non_jokers[4]:
                    return 'one_pair'
                else:
                    return 'card'

    def __lt__(self, other):
        if self.type_order.index(self.hand_type) < self.type_order.index(other.hand_type):
            return True
        if self.type_order.index(self.hand_type) > self.type_order.index(other.hand_type):
            return False

        for n in range(len(self.cards)):
            if self.card_order.index(self.cards[n]) < self.card_order.index(other.cards[n]):
                return True
            if self.card_order.index(self.cards[n]) > self.card_order.index(other.cards[n]):
                return False

        return False

    def __str__(self):
        return f"<{self.__class__.__name__}: {self.cards}, {self.bid}: {self.hand_type}>"

    def __repr__(self):
        return str(self)


lines = [row.strip() for row in open('07.input').readlines()]

hands = []
for line in lines:
    hands.append(Hand(line.split()[0], int(line.split()[1])))

hands.sort()

winnings = 0
for i in range(len(hands)):
    winnings += hands[i].bid * (i + 1)

print(winnings)
